csharp/src/Ice/OutputStream.cs

```diff
@@ -72,143 +67,102 @@ public Position(int segment, int offset)
-        /// </summary>
-        public void StartEncapsulation() => StartEncapsulation(Encoding);
-
-        /// <summary>
-        /// Writes the start of an encapsulation to the stream.
-        /// </summary>
-        /// <param name="encoding">The encoding version of the encapsulation.</param>
-        /// <param name="format">Specify the compact or sliced format; when null, keep the stream's format.</param>
-        public void StartEncapsulation(Encoding encoding, FormatType? format = null)
-        {
-            Debug.Assert(_mainEncaps == null && _endpointEncaps == null);
-            encoding.CheckSupported();
-
-            _mainEncaps = new Encaps(Encoding, _format, _tail);
-
-            Encoding = encoding;
-            if (format.HasValue)
+            if (startPayload)
             {
-                _format = format.Value;
-            }
-
-            WriteEncapsulationHeader(0, Encoding);
-        }
+                Debug.Assert(payloadReady != null);
+                _payloadReady = payloadReady;
+                _mainEncaps = new Encaps(Encoding, _tail);
 
-        internal IList<ArraySegment<byte>> GetUnderlyingBuffer()
-        {
-            Debug.Assert(_segmentList.Count > 0);
-            var buffer = new List<ArraySegment<byte>>();
-            for (int i = 0; i < _segmentList.Count; i++)
-            {
-                ArraySegment<byte> segment = _segmentList[i];
-                if (i == _tail.Segment)
+                if (format.HasValue)
                 {
-                    buffer.Add(segment.Slice(0, _tail.Offset));
-                    break;
-                }
-                else
-                {
-                    buffer.Add(segment);
+                    _format = format.Value;
                 }
+
+                WriteEncapsulationHeader(0, Encoding);
             }
-            return buffer;
         }
 
-        /// <summary>
-        /// Ends the previous main encapsulation.
-        /// </summary>
-        public void EndEncapsulation()
+        public Position Save()
         {
-            Debug.Assert(_mainEncaps.HasValue && _endpointEncaps == null);
+            _segmentList[_tail.Segment] = _segmentList[_tail.Segment].Slice(0, _tail.Offset);
 
-            Encaps encaps = _mainEncaps.Value;
+            if (_mainEncaps.HasValue)
+            {
+                Debug.Assert(_endpointEncaps == null);
 
-            // Size includes size and version.
-            RewriteInt(Distance(encaps.StartPos), encaps.StartPos);
+                Encaps encaps = _mainEncaps.Value;
 
-            Encoding = encaps.OldEncoding;
-            _format = encaps.OldFormat;
-            ResetEncapsulation();
+                // Size includes size and version.
+                RewriteInt(Distance(encaps.StartPos), encaps.StartPos);
+
+                Debug.Assert(_payloadReady != null);
+
+                _payloadReady(_tail);
+            }
+            return _tail;
```

Comment:
I think once Saved the OutputStream should no longer be used, but I don't see an easy way to do that. For example, can't set _segmentList to null since it's readonly. Add a TODO to figure this out later?